### PR TITLE
Parser improvements

### PR DESCRIPTION
**Pull request by trowski**

### CODE REVIEW

File: src/BufferingParser.php [hunk 1]
```diff
@@ -63,20 +64,19 @@ private function parseBody(string $body, string $boundary = null): Form
...
 
+            foreach (\explode("&", $body, $this->fieldCountLimit) as $pair) {
                 $pair = \explode("=", $pair, 2);
                 $field = \urldecode($pair[0]);
                 $value = \urldecode($pair[1] ?? "");
 
                 $fields[$field][] = $value;
             }
 
+            if (\strpos($value, "&") !== false) {
```
Comment (kelunik) [inline src/BufferingParser.php]: `$value` might be undefined here.

File: src/BufferingParser.php [hunk 2]
```diff
@@ -87,44 +87,49 @@ private function parseBody(string $body, string $boundary = null): Form
...
-            $headers = [];
-
-            foreach (\explode("\r\n", $rawHeaders) as $header) {
-                $split = \explode(":", $header, 2);
-                if (!isset($split[1])) {
-                    return new Form([]);
-                }
-                $headers[\strtolower($split[0])] = \trim($split[1]);
+            if (($position = \strpos($entry, "\r\n\r\n")) === false) {
+                throw new ParseException("No header/body boundry found");
```
Comment (kelunik) [inline src/BufferingParser.php]: `boundry` → `boundary`

File: src/BufferingParser.php [hunk 3]
```diff
@@ -87,44 +87,49 @@ private function parseBody(string $body, string $boundary = null): Form
...
             $count = \preg_match(
                 '#^\s*form-data(?:\s*;\s*(?:name\s*=\s*"([^"]+)"|filename\s*=\s*"([^"]+)"))+\s*$#',
-                $headers["content-disposition"] ?? "",
+                $headers["content-disposition"][0] ?? "",
                 $matches
             );
 
             if (!$count || !isset($matches[1])) {
-                return new Form([]);
+                throw new ParseException("Missing or invalid content disposition");
```
Comment (kelunik) [inline src/BufferingParser.php:114]: This would be the only minor "BC break", right?

Comment (trowski) [reply src/BufferingParser.php:114]: Yes, it makes `BufferingParser` consistent with `StreamingParser`.

File: src/BufferingParser.php [hunk 4]
```diff
@@ -87,44 +87,49 @@ private function parseBody(string $body, string $boundary = null): Form
...
-            $headers = [];
-
-            foreach (\explode("\r\n", $rawHeaders) as $header) {
-                $split = \explode(":", $header, 2);
-                if (!isset($split[1])) {
-                    return new Form([]);
-                }
-                $headers[\strtolower($split[0])] = \trim($split[1]);
+            if (($position = \strpos($entry, "\r\n\r\n")) === false) {
+                throw new ParseException("No header/body boundary found");
```
Comment (trowski) [inline src/BufferingParser.php:96]: This is a minor BC break, though prior a warning was produced if `\r\n\r\n` was not found.

Comment (kelunik) [reply src/BufferingParser.php:96]: If a warning was previously produced, I think it's fine.

File: src/BufferingParser.php (hunk 3 above)
Comment (kelunik) [reply src/BufferingParser.php:114]: I think it's fine.

**Pull request merged**